here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/query/WhereClauseBuilder.kt

```diff
@@ -160,6 +163,81 @@ class WhereClauseBuilder(private val request: ReadFeatures) {
         }
     }
 
+    private fun whereTags() {
+        val tagQuery = request.query.tags
+        if(tagQuery != null){
+            if (where.isNotEmpty()) {
+                where.append(" AND (")
+            } else {
+                where.append(" (")
+            }
+            whereNestedTags(tagQuery)
+            where.append(")")
+        }
+    }
+
+    private fun whereNestedTags(tagQuery: ITagQuery) {
+        when (tagQuery) {
+            is TagNot -> not(tagQuery.query, this::whereNestedTags)
+            is TagOr -> or(tagQuery.filterNotNull(), this::whereNestedTags)
+            is TagAnd -> and(tagQuery.filterNotNull(), this::whereNestedTags)
+            is TagQuery -> resolveSingleTagQuery(tagQuery)
+        }
+    }
+
+    private fun resolveSingleTagQuery(tagQuery: TagQuery) {
+        when (tagQuery) {
+            is TagExists -> {
+                where.append("$tagsAsJsonb ?? '${tagQuery.name}'")
+            }
+
+            is TagValueIsNull -> {
+                where.append("${tagValue(tagQuery)} = null")
+            }
+
+            is TagValueIsBool -> {
+                if (tagQuery.value) {
+                    where.append(tagValue(tagQuery, PgType.BOOLEAN))
+                } else {
+                    where.append("not(${tagValue(tagQuery, PgType.BOOLEAN)}})")
+                }
+            }
+
+            is TagValueIsDouble -> {
+                val valuePlaceholder = placeholderForArg(tagQuery.value, PgType.DOUBLE)
+                val doubleOp = resolveDoubleOp(
+                    tagQuery.op,
+                    tagValue(tagQuery, PgType.DOUBLE),
+                    valuePlaceholder
+                )
+                where.append(doubleOp)
+            }
+
+            is TagValueIsString -> {
+                val valuePlaceholder = placeholderForArg(tagQuery.value, PgType.STRING)
+                val stringEquals = resolveStringOp(
+                    StringOp.EQUALS,
+                    tagValue(tagQuery, PgType.STRING),
+                    valuePlaceholder
+                )
+                where.append(stringEquals)
+            }
+
+            is TagValueMatches -> {
+                val regex = tagQuery.regex
+                where.append("$tagsAsJsonb @?? '\$.${tagQuery.name} ? (@ like_regex \"${regex}\")'")
+            }
+        }
+    }
+
+    private fun tagValue(tagQuery: TagQuery, castTo: PgType? = null): String {
+        return when (castTo) {
+            null -> "$tagsAsJsonb->'${tagQuery.name}'"
+            PgType.STRING -> "$tagsAsJsonb->>'${tagQuery.name}'"
+            else -> "($tagsAsJsonb->'${tagQuery.name}')::${castTo.value}"
```

Comment:
All of this prone to SQL injections and ASCII zero attacks.

Reply:
Wrapped this in prepared statement as so:
```
private fun selectTagValue(tagQuery: TagQuery, castTo: PgType? = null): String {
    val tagKeyPlaceholder = placeholderForArg(tagQuery.name, PgType.STRING)
    return when (castTo) {
        null -> "$tagsAsJsonb->$tagKeyPlaceholder"
        PgType.STRING -> "$tagsAsJsonb->>$tagKeyPlaceholder"
        else -> "($tagsAsJsonb->$tagKeyPlaceholder)::${castTo.value}"
    }
}
```

The tag key will be wrapped as `$N` and replaced in prepared statement.
The `tagsAsJsonb` alone is safe:
```
companion object {
    private val tagsAsJsonb = "naksha_tags(${PgColumn.tags}, ${PgColumn.flags})"
}```